man/man1/cap2any.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "CAP2ANY" "1" "December 05, 2021" "cap2any 4.0" "User Manual"
.hy
.SH NAME
.PP
cap2any - Capture from media device and save audio/video in any format
.SH SYNOPSIS
.PP
\f[B]cap2any\f[R] [ \f[B]-a\f[R] audio codec ] [ \f[B]-v\f[R] video
codec ] [ \f[B]-c\f[R] codec ] [ \f[B]-p\f[R] preset ] [ \f[B]-V\f[R]
screen device index ] [ \f[B]-A\f[R] audio device index ] [ \f[B]-o\f[R]
output format (lower case) ] [ \f[B]-O\f[R] output filename no suffix ]
[ \f[B]-q\f[R] scale ] [ \f[B]-r\f[R] rate ] [ \f[B]-b\f[R] audio
bitrate ] [ \f[B]-z\f[R] video bitrate ] [ \f[B]-s\f[R] size ] [
\f[B]-t\f[R] threads ] [ \f[B]-I\f[R] ] [ \f[B]-C\f[R] ] [ \f[B]-d\f[R]
] [ \f[B]-D\f[R] ] [ \f[B]-y\f[R] ] [ \f[B]-u\f[R] ]
.SH DESCRIPTION
.PP
Captures, converts, and saves audio and video from a media device.
Input device and output format can be specified on the command line.
Convenience links can be created to simplify capture to various formats.
When invoked as the link \f[I]cap2yyy\f[R], \f[B]cap2any\f[R] converts
and saves the capture in media format \f[I]YYY\f[R].
.PP
The convenience links to \f[I]cap2any\f[R] that are installed by default
are:
.IP \[bu] 2
cap2m4v - Capture and convert to M4V
.IP \[bu] 2
cap2mp4 - Capture and convert to MP4
.PP
Additional convenience links can be created by creating a new symbolic
link to \f[I]cap2any\f[R].
For example, to create a convenience link that captures and converts to
AVI video format, issue the following commands:
.IP
.nf
\f[C]
cd /usr/local/bin
sudo ln -s /usr/local/DoctorFreeScript/bin/cap2any cap2avi
\f[R]
.fi
.SH COMMAND LINE OPTIONS
.TP
\f[B]-d\f[R]
indicates tell me what you would do
.TP
\f[B]-C\f[R]
indicates use the character interface, ignore AppleScript
.TP
\f[B]-D\f[R]
indicates use default audio and video devices, don\[aq]t prompt
.TP
\f[B]-o output format\f[R]
specifies the 3 letter lower case output format
.TP
\f[B]-O output filename\f[R]
specifies the output filename without suffix
.TP
\f[B]-a audio codec\f[R]
specifies the output audio codec
.TP
\f[B]-v video codec\f[R]
specifies the output video codec
.TP
\f[B]-s size\f[R]
specifies the output video size (widthxheight)
.TP
\f[B]-c codec\f[R]
specifies the output codec
.TP
\f[B]-I\f[R]
indicates add the captured file to Apple Music
.TP
\f[B]-p preset\f[R]
specifies the ffmpeg preset to use Useful presets: ultrafast superfast
veryfast faster fast medium slow slower veryslow.
Default preset is \[aq]slow\[aq]
.TP
\f[B]-q scale\f[R]
specifies the qscale variable bit rate quality
.TP
\f[B]-r rate\f[R]
specifies the rate for Constant Rate Factor (CRF) encoding.
Use \[dq]-r 0\[dq] to disable for formats other than x264
.TP
\f[B]-b bitrate\f[R]
specifies the bitrate (default 128k)
.TP
\f[B]-t threads\f[R]
specifies the number of threads to use
.TP
\f[B]-y\f[R]
indicates overwrite output files without asking
.TP
\f[B]-u\f[R]
displays this usage message
.PP
Current invocation defaults to:
.PP
\f[V]/usr/local/bin/ffmpeg -f avfoundation -i :   -loglevel warning -strict -2 -vsync 2  /Users/doctorwhen/Movies/ScreenCaps/MP4/capture_40163.mp4\f[R]
.SH EXAMPLES
.TP
\f[B]cap2any\f[R]
The most common use is capture using the default parameters
.TP
\f[B]cap2any -s 640x480\f[R]
Specify output video size of 640 width by 480 height
.TP
\f[B]cap2any -v libx264 -p ipod640\f[R]
Specify libx264 output video codec and use ipod640 preset
.TP
\f[B]cap2any -b 512k -I\f[R]
Specify 512k bitrate and add capture file to Apple Music
.TP
\f[B]cap2any -v libx264 -r 0 -q 3\f[R]
Specify libx264 output video codec and variable bit rate
.SH REQUIREMENTS
.PP
This command is basically a wrapper script for \f[B]ffmpeg\f[R].
To install \f[B]ffmpeg\f[R], issue the following command:
.TP
\f[B]Debian Linux\f[R]
sudo apt install ffmpeg
.TP
\f[B]Fedora/CentOS/SuSE Linux\f[R]
sudo yum install ffmpeg
.TP
\f[B]Mac OS X\f[R]
brew install ffmpeg
.SH AUTHORS
.PP
Written by Ronald Record <<github@ronrecord.com>>
.SH LICENSING
.PP
CAP2ANY is distributed under an Open Source license.
See the file LICENSE in the CAP2ANY source distribution for information
on terms & conditions for accessing and otherwise using CAP2ANY and for
a DISCLAIMER OF ALL WARRANTIES.
.SH BUGS
.PP
Submit bug reports online at:
<<https://gitlab.com/doctorfree/DoctorFreeScripts/issues>>
.SH SEE ALSO
.PP
\f[B]cap2m4v\f[R](1), \f[B]cap2mp4\f[R](1)
.PP
Full documentation and sources at:
<<https://gitlab.com/doctorfree/DoctorFreeScripts>>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

Jan 27, 2022 Version 4.0.1 Release 2
    Minor release update with improved support for SD management utilities,
    extending supported platforms to Linux for the SD utils.

Jan 17, 2022 Version 4.0.1 Release 1
    Release update to Version 4.0.1 including additional man pages
    and Zip format distribution archive. Minor bug fixes and improvements.
    RPM installation fixes.

Nov 26, 2021 Version 4.0 Release 1
    This release adds Debian and RPM format installation package support with
    improved support for ease of installation, configuration, update, and removal.

Wed Jan 26 14:27:43 2022 -0800 b752940 :
   Exec /bin/bash in systemd service to comply with Fedora restrictions
Tue Jan 25 08:37:12 2022 -0800 f38080d :
   Add journalclean command
Tue Jan 25 07:14:52 2022 -0800 d4b92fc :
   Set backgrounds folder to /media/user/Transcend by default on Linux
Mon Jan 24 18:21:57 2022 -0800 363a940 :
   Change default SD device to /dev/sdd
Mon Jan 24 10:17:35 2022 -0800 3be3fab :
   Add documentation for sdmount and sdumount
Sun Jan 23 19:48:26 2022 -0800 4088f8d :
   Add sdmount and sdumount commands
Sun Jan 23 10:31:20 2022 -0800 68b537d :
   Doc for sdlist
Sun Jan 23 10:04:48 2022 -0800 04f8708 :
   Add sdlist command
Sat Jan 22 16:03:14 2022 -0800 02729e7 :
   Extend SD utilities with Linux support
Fri Jan 21 07:50:47 2022 -0800 6be5d58 :
   Muriel aliases
Thu Jan 20 11:45:13 2022 -0800 24d045d :
   Redirect stderr and stdout to parent process during RPM scriptlets
Tue Jan 18 17:17:31 2022 -0800 2152b2d :
   Use DISPLAY env var if set
Tue Jan 18 13:20:58 2022 -0800 85da437 :
   Add doctorinfo script
Mon Jan 17 13:52:51 2022 -0800 65cd49e :
   Adopt release creation during Gitlab CI
Mon Jan 17 13:42:38 2022 -0800 735ff4e :
   Do not invoke dl_fred_scripts during postinstall, new version string
Mon Jan 17 12:06:57 2022 -0800 6bc24c5 :
   Move pvoutput to MirrorCommand
Mon Jan 17 12:06:31 2022 -0800 8dd49ed :
   Bump release
Fri Jan 14 13:52:28 2022 -0800 1ab3135 :
   Update miner scripts
Fri Jan 14 10:49:45 2022 -0800 988389c :
   Add disable-update-notifier and pkg/tools/get_release
Mon Jan 10 12:23:00 2022 -0800 6e4799a :
   Do not package or export gdrive utils
Mon Jan 10 12:19:41 2022 -0800 9c85af0 :
   Remove gdrive files as they are now packaged in their own repo
Sun Jan 9 17:43:49 2022 -0800 8962417 :
   disable rpmbuild strip
Sun Jan 9 17:18:32 2022 -0800 5af1d70 :
   Fix files section of spec file
Sun Jan 9 17:10:20 2022 -0800 905d402 :
   Move all files into /usr/local
Sun Jan 9 16:39:42 2022 -0800 e27d740 :
   Try noninteractive rpm install again
Sun Jan 9 15:35:04 2022 -0800 64ddc3a :
   Fix for CI
Sun Jan 9 15:28:21 2022 -0800 fd5915b :
   export noninteractive
Sun Jan 9 15:18:00 2022 -0800 c6e51e9 :
   Copy rpm dist tree in install phase
Sun Jan 9 14:46:58 2022 -0800 bfa4e07 :
   Install rpm in CI in noninteractive mode with TZ set
Sun Jan 9 14:17:31 2022 -0800 53255f8 :
   Use ubuntu:latest CI image
Sun Jan 9 14:09:05 2022 -0800 c6963e9 :
   Version 4.0 release 3
Sun Jan 9 14:04:52 2022 -0800 fd33ea1 :
   Install zip if not found, create zip artifacts in CI
Thu Jan 6 11:57:39 2022 -0800 3b3f004 :
   Add drive to projects
Mon Jan 3 20:15:25 2022 -0800 748b3e0 :
   Add author attribution
Mon Jan 3 16:22:13 2022 -0800 236f864 :
   Add -l option to list sync folders
Mon Jan 3 16:11:26 2022 -0800 d30988a :
   Renamed getfileids as getfolderids
Mon Jan 3 13:01:47 2022 -0800 97d0e8d :
   Check for gdrive binary in PATH
Mon Jan 3 12:38:50 2022 -0800 ded8f3d :
   Add options to getfolderids, start on support for spaces in file and folder names
Mon Jan 3 10:24:36 2022 -0800 627a90d :
   Support spaces in folder names, retrieve all folder ids
Sun Jan 2 21:10:06 2022 -0800 1ec7cb9 :
   Export gdrive utils
Sun Jan 2 21:07:06 2022 -0800 40e2bdb :
   Add gdrive scripts to exports
Sun Jan 2 21:04:30 2022 -0800 e9d2a92 :
   Add comment for developer's guide to api
Sun Jan 2 20:19:56 2022 -0800 c674bc0 :
   Add sync_from_drive, update sync2drive
Sun Jan 2 11:48:46 2022 -0800 285acd8 :
   Updated gdrive utility scripts
Sat Jan 1 20:14:40 2022 -0800 a670a0d :
   Add new scripts for command line management of Google Drive
Fri Dec 31 08:52:06 2021 -0800 a1659d2 :
   Add Mirco Cabbia alias
Tue Dec 28 09:50:38 2021 -0800 904ab6d :
   Rename MirrorCommandLine to MirrorCommand
Mon Dec 27 18:20:34 2021 -0800 3554d70 :
   New model alias
Mon Dec 27 08:43:35 2021 -0800 5a2a524 :
   New model alias
Mon Dec 27 08:23:33 2021 -0800 a981875 :
   Create man page for md2man
Sun Dec 26 17:48:24 2021 -0800 6d83be7 :
   Set output dir based on man page section in md2man
Sat Dec 25 08:04:35 2021 -0800 ef79abb :
   Added Colombian wallpapers
Fri Dec 24 11:34:02 2021 -0800 0c9e9cc :
   Adjusted aliases
Thu Dec 23 09:38:25 2021 -0800 c2b5191 :
   Updated model name, add Mila Azul to Ukrainians
Wed Dec 22 16:31:35 2021 -0800 d4296b4 :
   Use HEAD to determine branch
Tue Dec 21 14:01:20 2021 -0800 526124a :
   Bump release number, update changelog
Tue Dec 21 10:26:02 2021 -0800 8c3a636 :
   Add all my clone and build scripts
Tue Dec 21 09:43:04 2021 -0800 d625a49 :
   Check and set location of Synergy client binary and system tools in Synergy startup control script
Tue Dec 21 09:14:43 2021 -0800 7a101e1 :
   Add man page for sdwrite, add SEE ALSO pages to all SD card commands
Tue Dec 21 08:40:29 2021 -0800 58b2dfc :
   Move sdwrite from MirrorCommandLine package to here
Sat Dec 18 17:18:51 2021 -0800 148f60a :
   Set synergyctl IP in postinstall
Sat Dec 18 16:39:57 2021 -0800 a8df6ce :
   Add synergyctl script to control Synergy start/stop
Fri Dec 17 07:15:45 2021 -0800 30fd5e5 :
   New model aliases
Mon Dec 13 21:11:37 2021 -0800 2c51ae3 :
   Fix typo in rmbrokenpkg
Mon Dec 13 20:18:02 2021 -0800 e5436b5 :
   Add rmbrokenpkg
Sun Dec 12 13:40:26 2021 -0800 2a4714d :

scripts/cap2any.sh
#!/bin/bash
#
## @file cap2any.sh
## @brief Capture the screen and output to the specified audio/video format
## @author Ronald Joe Record (rr at ronrecord dot com)
## @copyright Copyright (c) 2015, Ronald Joe Record, all rights reserved.
## @date Written 6-Jul-2015
## @version 1.0.1
##
## See https://trac.ffmpeg.org/wiki/Capture/Desktop
##
## Use the avfoundation device:
##
## ffmpeg -f avfoundation -list_devices true -i ""
##
## This will enumerate all the available input devices including
## screens ready to be captured. Once you've figured out the device
## index corresponding to the screen to be captured use:
##
## ffmpeg -f avfoundation -i "<scr device index>:<aud device index>" out.m4v
##
## This will capture the screen from <screen device index> and audio from
## <audio device index> into the output file out.m4v.
##
## Uses the default options for audio/video encoding from my any2any script
##
##    This program works by either linking or copying cap2any to a file
##    which specifies the desired output format by its name.
##    Alternately, the -o command line option can be used to specify
##    the output file format.
##
##    For example, if you want to capture to MP4 then you could
##    create a symbolic link from cap2any to cap2mp4 as follows:
##        ln -s cap2any cap2mp4
##    Similarly, symbolic links (or copies or hard links) could be created to
##    capture to any other audio/video format.
##
##    Naming restricton: cap2[3 lowercase letter output]
##    for a 7 letter name beginning with "cap2". The 3 letter prefix and
##    suffix must also be a filename suffix that ffmpeg recognizes as a valid
##    audio/video format.
##
##    For example, the following are filenames that conform to this restriction:
##        cap2mp4 cap2mpg cap2mkv cap2m4a cap2wmv
##
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# The Software is provided "as is", without warranty of any kind, express or
# implied, including but not limited to the warranties of merchantability,
# fitness for a particular purpose and noninfringement. In no event shall the
# authors or copyright holders be liable for any claim, damages or other
# liability, whether in an action of contract, tort or otherwise, arising from,
# out of or in connection with the Software or the use or other dealings in
# the Software.
#
## AVFoundation supports the following options:
## 
## -list_devices <TRUE|FALSE>
##    If set to true, a list of all available input devices is given showing
##    all device names and indices.
## -video_device_index <INDEX>
##    Specify the video device by its index. Overrides anything given in the
##    input filename.
## -audio_device_index <INDEX>
##    Specify the audio device by its index. Overrides anything given in the
##    input filename.
## -pixel_format <FORMAT>
##    Request the video device to use a specific pixel format. If the specified
##    format is not supported a list of available formats is given and the first
##    one in this list is used instead. Available pixel formats are: monob,
##    rgb555be, rgb555le, rgb565be, rgb565le, rgb24, bgr24, 0rgb, bgr0, 0bgr,
##    rgb0, bgr48be, uyvy422, yuva444p, yuva444p16le, yuv444p, yuv422p16,
##    yuv422p10, yuv444p10, yuv420p, nv12, yuyv422, gray
## 
## Examples
## 
## Print the list of AVFoundation supported devices and exit:
## $ ffmpeg -f avfoundation -list_devices true -i ""
## 
## Record video from video device 0 and audio from audio device 0 into out.avi:
## $ ffmpeg -f avfoundation -i "0:0" out.avi
## 
## Record video from video device 2 and audio from audio device 1 into out.avi:
## $ ffmpeg -f avfoundation -video_device_index 2 -i ":1" out.avi
## 
## Record video from the system default video device using the pixel format bgr0
## and do not record any audio into out.avi:
## $ ffmpeg -f avfoundation -pixel_format bgr0 -i "default:none" out.avi
## 

## @fn GetVideoNum()
## @brief Lists devices and retrieves video device index
## @param none
##
## Uses AppleScript, if available, to provide graphical user interface
GetVideoNum() {
    video_devs=
    video_nums=
    ${FFMPEG} -f avfoundation -list_devices true -i "" 2>&1 | grep AVFoundation | grep -v "devices:" | egrep -i "Video|Camera|Screen" | awk ' { for(i=6;i<NF;i++) printf "%s",$i OFS; if (NF) printf "%s",$NF; printf ORS}' > /tmp/__avout$$
    while read line
    do
        devnam=`echo $line | awk ' { for(i=2;i<NF;i++) printf "%s",$i OFS; if (NF) printf "%s",$NF; printf ORS } '`
        devnum=`echo $line | awk ' { print $1 } '`
        if [ "$video_devs" ]
        then
            video_devs="$video_devs, \"$devnam\""
            video_nums="$video_nums, \"$devnum\""
            video_dev_names=("${video_dev_names[@]}" "$devnam")
            video_dev_numbs=("${video_dev_numbs[@]}" "$devnum")
        else
            video_devs="\"$devnam\""
            video_nums="\"$devnum\""
            video_dev_names=("$devnam")
            video_dev_numbs=("$devnum")
        fi
    done < /tmp/__avout$$
    video_devs="$video_devs, \"default\", \"none\""
    video_nums="$video_nums, \"default\", \"none\""
    video_dev_names=("${video_dev_names[@]}" "default" "none")
    video_dev_numbs=("${video_dev_numbs[@]}" "default" "none")

    if [ "$APPLE_SCRIPT" ]
    then
        selected_video_device=$(osascript <<EOF
            set video_devices to {$video_devs}
            set video_numbers to {$video_nums}
            set video_device to (choose from list video_devices with title "AVFoundation Video Devices" with prompt "Select a video device" default items {"$default_video_device"})
    
            set video_number to "$default_video_number"
            if video_device is false then
                set video_device to "$default_video_device"
            else
                set video_number to item (list_position((video_device as string), video_devices)) of video_numbers
            end if
            return video_number as string
    
            on list_position(this_item, this_list)
                repeat with i from 1 to count of this_list
                    if item i of this_list is this_item then return i
                end repeat
                return 0
            end list_position
EOF)
        VIDEO=`echo ${selected_video_device} | sed -e "s/\[//" -e "s/\]//"`
    else
        # Use the Bash select statement to get user device selection
        video_num_size=${#video_dev_numbs[@]}
        printf '\nSelect the video device you wish to capture:\n\n'
        PS3='Video device number? '
        select device in "${video_dev_names[@]}"
        do
            if ([ $REPLY -gt 0 ] && [ $REPLY -le $video_num_size ])
            then
                video_index=`expr $REPLY - 1`
                video_device=${video_dev_numbs[$video_index]}
                VIDEO=`echo ${video_device} | sed -e "s/\[//" -e "s/\]//"`
                break
            else
                printf "Unexpected reply: $REPLY"
            fi
        done
    fi
}

## @fn GetAudeoNum()
## @brief Lists devices and retrieves audio device index
## @param none
##
## Uses AppleScript, if available, to provide graphical user interface
GetAudioNum() {
    audio_devs=
    audio_nums=
    ${FFMPEG} -f avfoundation -list_devices true -i "" 2>&1 | grep AVFoundation | grep -v "devices:" | egrep -i "Audio|Sound|Microphone" | awk ' { for(i=6;i<NF;i++) printf "%s",$i OFS; if (NF) printf "%s",$NF; printf ORS}' > /tmp/__avout$$
    while read line
    do
        devnam=`echo $line | awk ' { for(i=2;i<NF;i++) printf "%s",$i OFS; if (NF) printf "%s",$NF; printf ORS } '`
        devnum=`echo $line | awk ' { print $1 } '`
        if [ "$audio_devs" ]
        then
            audio_devs="$audio_devs, \"$devnam\""
            audio_nums="$audio_nums, \"$devnum\""
            audio_dev_names=("${audio_dev_names[@]}" "$devnam")
            audio_dev_numbs=("${audio_dev_numbs[@]}" "$devnum")
        else
            audio_devs="\"$devnam\""
            audio_nums="\"$devnum\""
            audio_dev_names=("$devnam")
            audio_dev_numbs=("$devnum")
        fi
    done < /tmp/__avout$$
    audio_devs="$audio_devs, \"default\", \"none\""
    audio_nums="$audio_nums, \"default\", \"none\""
    audio_dev_names=("${audio_dev_names[@]}" "default" "none")
    audio_dev_numbs=("${audio_dev_numbs[@]}" "default" "none")

    if [ "$APPLE_SCRIPT" ]
    then
        selected_audio_device=$(osascript <<EOF
            set audio_devices to {$audio_devs}
            set audio_numbers to {$audio_nums}
            set audio_device to (choose from list audio_devices with title "AVFoundation Audio Devices" with prompt "Select an audio device" default items {"$default_audio_device"})
    
            if audio_device is false then
                set audio_device to "$default_audio_device"
                set audio_number to "$default_audio_number"
            else
                set audio_number to item (list_position((audio_device as string), audio_devices)) of audio_numbers
            end if
            return audio_number as string
    
            on list_position(this_item, this_list)
                repeat with i from 1 to count of this_list
                    if item i of this_list is this_item then return i
                end repeat
                return 0
            end list_position
EOF)
        AUDIO=`echo ${selected_audio_device} | sed -e "s/\[//" -e "s/\]//"`
    else
        # Use the Bash select statement to get user device selection
        audio_num_size=${#audio_dev_numbs[@]}
        printf '\nSelect the audio device you wish to capture:\n\n'
        PS3='Audio device number? '
        select device in "${audio_dev_names[@]}"
        do
            if ([ $REPLY -gt 0 ] && [ $REPLY -le $audio_num_size ])
            then
                audio_index=`expr $REPLY - 1`
                audio_device=${audio_dev_numbs[$audio_index]}
                AUDIO=`echo ${audio_device} | sed -e "s/\[//" -e "s/\]//"`
                break
            else
                printf "Unexpected reply: $REPLY"
            fi
        done
    fi
}

# I put the latest experimental ffmpeg in /usr/local/bin but if it's not there
# simply revert to whatever is first in our PATH
FFMPEG=/usr/local/bin/ffmpeg
[ -x ${FFMPEG} ] || FFMPEG=ffmpeg
prompt_video=1
prompt_audio=1
default_audio_device=`${FFMPEG} -f avfoundation -list_devices true -i "" 2>&1 | grep AVFoundation | grep Built | grep Microphone | head -1 | awk ' { for(i=7;i<NF;i++) printf "%s",$i OFS; if (NF) printf "%s",$NF; printf ORS}'`
default_audio_number=`${FFMPEG} -f avfoundation -list_devices true -i "" 2>&1 | grep AVFoundation | grep Built | grep Microphone | head -1 | awk ' {print $6}'`
default_video_device=`${FFMPEG} -f avfoundation -list_devices true -i "" 2>&1 | grep AVFoundation | grep Capture | grep screen | head -1 | awk ' { for(i=7;i<NF;i++) printf "%s",$i OFS; if (NF) printf "%s",$NF; printf ORS}'`
default_video_number=`${FFMPEG} -f avfoundation -list_devices true -i "" 2>&1 | grep AVFoundation | grep Capture | grep screen | head -1 | awk ' { print $6 }'`

ME=`basename $0`
if [ "$ME" = "cap2any" ]
then
    SUF="mp4"
    DIR="MP4"
else
    SUF="${ME:4:3}"
    DIR=`echo $SUF | tr '[:lower:]' '[:upper:]'`
fi

CODEC=
SIZE=
OVER=
TELL=
USAGE=
ACODEC=
VCODEC=
OUT_PRE=
PRESET=
QSCALE=
RATE=
ASAMPLE=
VSAMPLE=
CHARACTER=
APPLE_SCRIPT=
THREAD=
ITUNES=

## @fn usage()
## @brief Display command line usage options
## @param none
##
## Exit the program after displaying the usage message and example invocations
usage() {
  printf "\nUsage: $ME [-a audio codec] [-v video codec] [-c codec] [-p preset]"
  printf "\n\t\t[-V screen device index] [-A audio device index]"
  printf "\n\t\t[-o output format (lower case)] [-O output filename no suffix]"
  printf "\n\t\t[-q scale] [-r rate] [-b audio bitrate] [-z video bitrate]"
  printf "\n\t\t[-s size] [-t threads] [-I] [-C] [-d] [-D] [-y] [-u]"
  printf "\n\nWhere:\n\t-d indicates tell me what you would do"
  printf "\n\t-C indicates use the character interface, ignore AppleScript"
  printf "\n\t-D indicates use default audio and video devices, don't prompt"
  printf "\n\t-o output format specifies the 3 letter lower case output format"
  printf "\n\t-O output filename specifies the output filename without suffix"
  printf "\n\t-a audio codec specifies the output audio codec"
  printf "\n\t-v video codec specifies the output video codec"
  printf "\n\t-s size specifies the output video size (widthxheight)"
  printf "\n\t-c codec specifies the output codec"
  printf "\n\t-I indicates add the captured file to Apple Music"
  printf "\n\t-p preset specifies the ffmpeg preset to use"
  printf "\n\t\t Useful presets:"
  printf "\n\t\t ultrafast superfast veryfast faster fast medium slow"
  printf "\n\t\t slower veryslow. Default preset is 'slow'"
  printf "\n\t-q scale specifies the qscale variable bit rate quality"
  printf "\n\t-r rate specifies the rate for Constant Rate Factor (CRF)"
  printf "\n\t\tencoding. Use \"-r 0\" to disable for formats other than x264"
  printf "\n\t-b bitrate specifies the bitrate (default 128k)"
  printf "\n\t-t threads specifies the number of threads to use"
  printf "\n\t-y indicates overwrite output files without asking"
  printf "\n\t-u displays this usage message\n"
  printf "\nCurrent invocation defaults to:\n\n${FFMPEG} ${CAP} ${OPTS} ${OUT_FILE}\n"
  printf "\nExamples:"
  printf "\n\tThe most common use is capture using the default parameters"
  printf "\n\t\t$ME"
  printf "\n\tSpecify output video size of 640 width by 480 height"
  printf "\n\t\t$ME -s 640x480"
  printf "\n\tSpecify libx264 output video codec and use ipod640 preset"
  printf "\n\t\t$ME -v libx264 -p ipod640"
  printf "\n\tSpecify 512k bitrate and add capture file to Apple Music"
  printf "\n\t\t$ME -b 512k -I"
  printf "\n\tSpecify libx264 output video codec and variable bit rate"
  printf "\n\t\t$ME -v libx264 -r 0 -q 3\n"
  exit 1
}

while getopts A:V:o:O:p:q:r:b:t:s:v:z:a:c:IydCDu flag; do
    case $flag in
        A)
            AUDIO="$OPTARG"
            prompt_audio=
            ;;
        V)
            VIDEO="$OPTARG"
            prompt_video=
            ;;
        a)
            ACODEC="-acodec $OPTARG"
            ;;
        b)
            ASAMPLE="-ab $OPTARG"
            ;;
        o)
            SUF=$OPTARG
            DIR=`echo $SUF | tr '[:lower:]' '[:upper:]'`
            ;;
        O)
            OUT_PRE=$OPTARG
            ;;
        z)
            VSAMPLE="-vb $OPTARG"
            ;;
        c)
            CODEC="-codec $OPTARG"
            ;;
        C)
            CHARACTER=1
            ;;
        d)
            TELL=1
            ;;
        D)
            prompt_audio=
            prompt_video=
            ;;
        I)
            ITUNES=1
            ;;
        p)
            PRESET="-preset $OPTARG"
            ;;
        q)
            QSCALE="-qscale $OPTARG"
            RATE=
            ;;
        r)
            if [ $OPTARG -eq 0 ]
            then
                RATE=
            else
                RATE="-crf $OPTARG"
            fi
            ;;
        s)
            SIZE="-s $OPTARG"
            ;;
        t)
            THREAD="-threads $OPTARG"
            ;;
        y)
            OVER="-y"
            ;;
        v)
            VCODEC="-vcodec $OPTARG"
            ;;
        u)
            USAGE=1
            ;;
    esac
done
shift $(( OPTIND - 1 ))

inst=`type -p osascript`
[ "$inst" ] && APPLE_SCRIPT=1
[ "$CHARACTER" ] && APPLE_SCRIPT=
if ([ "$APPLE_SCRIPT" ] && [ "$ITUNES" ])
then
    echo "Capture will automatically be imported into Apple Music"
else
    if [ "$ITUNES" ]
    then
        echo "AppleScript is not supported on this platform."
        echo "Unable to automate the installation in Apple Music."
        ITUNES=
    fi
fi
OUT_DIR=$HOME/Movies/ScreenCaps/$DIR
[ -d "${OUT_DIR}" ] || mkdir -p "${OUT_DIR}"
[ "${OUT_PRE}" ] || OUT_PRE="capture_$$"
OUT_FILE=${OUT_DIR}/${OUT_PRE}.$SUF
[ "$USAGE" ] || {
    [ "$prompt_audio" ] && GetAudioNum
    [ "$prompt_video" ] && GetVideoNum
}
[ "$AUDIO" ] || {
    AUDIO=`echo ${default_audio_number} | sed -e "s/\[//" -e "s/\]//"`
}
[ "$VIDEO" ] || {
    VIDEO=`echo ${default_video_number} | sed -e "s/\[//" -e "s/\]//"`
}
# Set ffmpeg capture options based upon the desired output format
LOG="-loglevel warning"
# RAT="-framerate 30"
RAT=
EXP="-strict -2"
# PIX="-pix_fmt yuv420p"
PIX=
VSY="-vsync 2"
#
# Set any options specified on the command line - the sed part is just
# removing any leading or trailing spaces.
OPTS=`echo "$OVER $ACODEC $CODEC $SIZE $VCODEC $PRESET $QSCALE $RATE $ASAMPLE $VSAMPLE $THREAD" | sed -e 's/^ *//' -e 's/ *$//'`
CAP="-f avfoundation -i $VIDEO:$AUDIO ${PIX} ${RAT} ${LOG} ${EXP} ${VSY}"
# Check for incompatible ffmpeg arguments
[ "$RATE" ] && [ "$QSCALE" ] && {
    printf "\nCannot specify both constant and variable bit rate values."
    printf "\nCurrent invocation would invoke:\n\n${FFMPEG} ${CAP} ${OPTS} ${OUT_FILE}\n"
    printf "\nExiting.\n"
    exit 1
}

# Display usage message and exit if -u was given on the command line.
# We delay until now in order to gather up all the other command line options.
[ "$USAGE" ] && usage

printf "\n${FFMPEG} ${CAP} ${OPTS} ${OUT_FILE}\n"
[ "$TELL" ] || {
    ${FFMPEG} ${CAP} ${OPTS} ${OUT_FILE}

    [ "$ITUNES" ] && {
        # Construct full pathname of output file
        k=`pwd`/"$j"
        echo "Adding $j to Apple Music"
        osascript -e "tell application \"Music\" to add POSIX file \"$k\""
    }
}
